cluster-autoscaler/cloudprovider/test/test_cloud_provider.go

```diff
@@ -0,0 +1,188 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package test
+
+import (
+	"fmt"
+	"sync"
+
+	"k8s.io/contrib/cluster-autoscaler/cloudprovider"
+	apiv1 "k8s.io/kubernetes/pkg/api/v1"
+)
+
+// TestCloudProvider is a dummy cloud provider to be used in tests.
+type TestCloudProvider struct {
+	sync.Mutex
+	nodes      map[string]string
+	groups     map[string]cloudprovider.NodeGroup
+	onIncrease func(string, int) error
+	onDelete   func(string, string) error
+}
+
+// OnIncreaseFunc is a function called on node group increase in TestCloudProvider.
+// First parameter is the NodeGroup id, second is the increase delta.
+type OnIncreaseFunc func(string, int) error
+
+// OnDeleteFunc is a function called on cluster
+type OnDeleteFunc func(string, string) error
+
+// NewTestCloudProvider builds new TestCloudProvider
+func NewTestCloudProvider(onIncrease OnIncreaseFunc, onDelete OnDeleteFunc) *TestCloudProvider {
+	return &TestCloudProvider{
+		nodes:      make(map[string]string),
+		groups:     make(map[string]cloudprovider.NodeGroup),
+		onIncrease: onIncrease,
+		onDelete:   onDelete,
+	}
+}
+
+// Name returns name of the cloud provider.
+func (tcp *TestCloudProvider) Name() string {
+	return "TestCloudProvider"
+}
+
+// NodeGroups returns all node groups configured for this cloud provider.
+func (tcp *TestCloudProvider) NodeGroups() []cloudprovider.NodeGroup {
+	tcp.Lock()
+	defer tcp.Unlock()
+
+	result := make([]cloudprovider.NodeGroup, 0)
+	for _, group := range tcp.groups {
+		result = append(result, group)
+	}
+	return result
+}
+
+// NodeGroupForNode returns the node group for the given node, nil if the node
+// should not be processed by cluster autoscaler, or non-nil error if such
+// occurred.
+func (tcp *TestCloudProvider) NodeGroupForNode(node *apiv1.Node) (cloudprovider.NodeGroup, error) {
```

Comment:
This function currently never returns error, so, I'm not sure if it is reasonable to keep error in the signature.